File: apps/api/app/wallet_detect.py
"""Detect wallet chain from address format."""

from __future__ import annotations

from typing import Literal, Optional

from .btc_fetch import is_valid_btc_address
from .cardano_fetch import is_valid_cardano_address
from .celestia_fetch import is_valid_celestia_address
from .evm_chains import EVM_CHAIN_META, EVM_AUTO_IMPORT_LABEL, EvmChain, is_evm_chain
from .evm_fetch import is_valid_evm_address
from .solana_fetch import is_valid_solana_address

WalletChain = Literal[
    "solana",
    "ethereum",
    "bitcoin",
    "cardano",
    "celestia",
    "arbitrum",
    "base",
    "optimism",
    "polygon",
    "bsc",
    "avalanche",
    "linea",
    "blast",
    "scroll",
]

CHAIN_LABELS: dict[str, str] = {
    "solana": "Solana",
    "ethereum": f"On-chain ({EVM_AUTO_IMPORT_LABEL}) + Hyperliquid",
    "bitcoin": "Bitcoin",
    "cardano": "Cardano",
    "celestia": "Celestia",
    **{slug: meta["label"] for slug, meta in EVM_CHAIN_META.items()},
}


def detect_wallet_chain(address: str) -> WalletChain:
    """Infer chain from address shape. Raises ValueError when unrecognized."""
    text = address.strip()
    if not text:
        raise ValueError("Wallet address is required.")

    lower = text.lower()

    if is_valid_evm_address(text):
        return "ethereum"

    if lower.startswith("celestia1"):
        if is_valid_celestia_address(lower):
            return "celestia"
        raise ValueError("Invalid Celestia address (expected celestia1…).")

    if lower.startswith("addr1") or lower.startswith("stake1"):
        if is_valid_cardano_address(lower):
            return "cardano"
        raise ValueError("Invalid Cardano address (expected addr1… or stake1…).")

    if is_valid_btc_address(text):
        return "bitcoin"

    if is_valid_solana_address(text):
        return "solana"

    raise ValueError(
        "Unrecognized wallet address. Supported formats: 0x… (EVM), Solana, "
        "Bitcoin, Cardano (addr1…), Celestia (celestia1…)."
    )


def normalize_wallet_address(address: str, chain: str) -> str:
    if chain in {"cardano", "celestia"}:
        return address.strip().lower()
    return address.strip()


def resolve_wallet_import(
    address: str,
    chain: Optional[WalletChain] = None,
) -> tuple[str, WalletChain]:
    """Return normalized address and chain (explicit chain must match address)."""
    text = address.strip()
    detected = detect_wallet_chain(text)
    chosen: WalletChain = chain or detected

    if chain is not None and chain != detected:
        if is_evm_chain(chain) and detected == "ethereum":
            chosen = chain
        else:
            raise ValueError(
                f"Address looks like {CHAIN_LABELS[detected]}, not {CHAIN_LABELS.get(chain, chain)}."
            )

    normalized = normalize_wallet_address(text, chosen)
    return normalized, chosen

File: apps/web/src/lib/walletDetect.ts
import type { WalletChain } from "@/lib/types";

/** Major on-chain networks for 0x import (matches backend evm_chains.py). */
export const EVM_AUTO_IMPORT_LABEL =
  "Ethereum, BNB Chain, Arbitrum, Base, Polygon, Optimism, Avalanche";

/** Ledger `source` values from on-chain or venue wallet import. */
export const WALLET_SOURCE_HINTS = new Set<string>([
  "solana",
  "bitcoin",
  "cardano",
  "celestia",
  "hyperliquid",
  "ethereum",
  "arbitrum",
  "base",
  "optimism",
  "polygon",
  "bsc",
  "avalanche",
  "linea",
  "blast",
  "scroll",
]);

export const WALLET_CHAIN_LABELS: Record<WalletChain, string> = {
  solana: "Solana",
  ethereum: `0x address → ${EVM_AUTO_IMPORT_LABEL} + Hyperliquid`,
  bitcoin: "Bitcoin",
  cardano: "Cardano",
  celestia: "Celestia",
};

const EVM_RE = /^0x[0-9a-fA-F]{40}$/;
const BTC_RE =
  /^(?:[13][a-km-zA-HJ-NP-Z1-9]{25,34}|bc1[ac-hj-np-z02-9]{11,71}|bc1p[ac-hm-np-z02-9]{58})$/;
const CARDANO_RE = /^(?:addr1[a-z0-9]{50,}|stake1[a-z0-9]{50,})$/;
const CELESTIA_RE = /^celestia1[a-z0-9]{38,}$/;
const SOLANA_RE = /^[1-9A-HJ-NP-Za-km-z]{32,44}$/;

/** Infer chain from address shape (matches backend wallet_detect). */
export function detectWalletChain(address: string): WalletChain | null {
  const text = address.trim();
  if (!text) return null;

  if (EVM_RE.test(text)) return "ethereum";

  const lower = text.toLowerCase();
  if (lower.startsWith("celestia1")) {
    return CELESTIA_RE.test(lower) ? "celestia" : null;
  }

  if (lower.startsWith("addr1") || lower.startsWith("stake1")) {
    return CARDANO_RE.test(lower) ? "cardano" : null;
  }

  if (BTC_RE.test(text)) return "bitcoin";

  if (SOLANA_RE.test(text)) return "solana";

  return null;
}

export function walletDetectError(address: string): string | null {
  const text = address.trim();
  if (!text) return null;
  if (detectWalletChain(text)) return null;
  return (
    "Unrecognized address — use Solana, 0x… (on-chain + Hyperliquid), Bitcoin, " +
    "Cardano (addr1…), or Celestia (celestia1…)"
  );
}
